File: leetcode/222-完全二叉树的节点个数/222-完全二叉树的节点个数.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */

// 222-完全二叉树的节点个数
var countNodes = function (root) {
    if (!root) { 
        return null;
    }
    let lDepth = 0,rDepth = 0;
    let l = root, r = root;
    while (l !== null) { 
        lDepth++;
        l = l.left;
    }
    
    while (r !== null) { 
        rDepth++;
        r = r.right;
    }
    if (lDepth === rDepth) { 
        return Number(Math.pow(2, lDepth) - 1);
    }
    return 1 + countNodes(root.left) + countNodes(root.right);
};

File: leetcode/222-完全二叉树的节点个数/222-完全二叉树的节点个数.py
import math

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

#  222-完全二叉树的节点个数
class Solution:
    def countNodes(self, root: Optional[TreeNode]) -> int:
        l,r = root,root
        l_depth,r_depth = 0,0
        while (not l is None):
            l_depth += 1
            l = l.left
        while (not r is None):
            r_depth += 1
            r = r.right
        if l_depth == r_depth:
            return int(math.pow(2,l_depth)-1)
        return 1 + self.countNodes(root.left) + self.countNodes(root.right)
